### DISCUSSION TITLE
OpenTelemetry for gRPC endpoints

### DISCUSSION THREAD

**Discussion by LarsSven** [quarkusio/quarkus#36220] [Q&A]
Hi all,

I have an application with a single gRPC endpoint. I want to use OpenTelemetry to trace the execution, and I am using the @WithSpan annotation to create spans within the trace. However, when I test it, I do get all the spans, but they're seen as separate:
![9VsfVvf](https://github.com/quarkusio/quarkus/assets/60571459/ed7109de-7e44-43db-a213-2877af81af34)
However I would like to see them as actual spans within my main trace. Is this a limitation of gRPC, or am I doing something wrong?

I have done no other configuration than installing the opentelemetry extension for Quarkus and adding @WithSpan to a few functions. From the documentation, I believe that is sufficient?

I hope someone can help me with this, thank you.

[a comment by an automated account is not shown]

**Comment by brunobat** [comment]
The gRPC server is already instrumented with OpenTelemetry. You only need to add `@WithSpan` on methods you create and need to track.
Please take a look at: https://github.com/quarkusio/quarkus/blob/main/integration-tests/opentelemetry-grpc/src/main/java/io/quarkus/it/opentelemetry/grpc/HelloService.java

Would you be able to provide a sample project that replicates the issue?

**Reply by LarsSven** [reply]
Thank you for the repsonse @brunobat,

I created a separate branch at https://gitlab.com/rug-digitallab/themis/judgement/companion-container/-/tree/ls-demo-otel?ref_type=heads. This branch has our current otel implementation added and has the features that make the program difficult to run disabled. You should be able to just run it pretty easily on any operating system through quarkusDev, as long as it has a docker daemon running. So make sure you run the `ls-demo-otel` branch.

Here is a file that you can run to make a request to the gRPC endpoint.
[GrpcClient.kt.md](https://github.com/quarkusio/quarkus/files/12775508/GrpcClient.kt.md)

If you run into any issues let me know and I can try to make a smaller program that demos the issue. I have never worked with opentelemetry before so it may be something I did wrong, but when I view the trace on jaeger, I just see 4 separate spans that don't seem to be connected.

**Reply by LarsSven** [reply]
(Keep in mind that the GrpcClient should be inside the project, as it uses some of the same dependencies. Also make sure you initialised the git submodules through `git submodule init --recursive`)

**Reply by cescoffier** [reply]
You are creating the client directly instead of using a client managed by Quarkus. Thus, the tracing interceptor is not used.

**Reply by LarsSven** [reply]
Ah okay, I did not know the request had to come from a Quarkus client.

Thank you, I will look into that.

**Reply by LarsSven** [reply]
So the issue was not related to that. It turns out that it was related to contexts not being automatically propagated through kotlin coroutines, so I need to manually propagate the context.
